db/version_set.cc

```diff
@@ -1577,11 +1587,42 @@ void LevelIterator::InitFileIterator(size_t new_file_index) {
       // no need to change anything
     } else {
       file_index_ = new_file_index;
-      InternalIterator* iter = NewFileIterator();
-      SetFileIterator(iter);
+      if (prepared_iters.size() > new_file_index &&
+          prepared_iters[new_file_index] != nullptr) {
+        SetFileIterator(prepared_iters[new_file_index]);
+      } else {
+        InternalIterator* iter = NewFileIterator(new_file_index);
+        SetFileIterator(iter);
+      }
     }
   }
 }
+
+void LevelIterator::PrepareFileIterator(size_t new_file_index) {
+  if (scan_opts_ != nullptr && scan_opts_->size() > 0) {
+    auto iter = NewFileIterator(new_file_index);
+    prepared_iters[new_file_index] = iter;
+    // We assume the scan_opts is ordered by start range.
+    const std::vector<ScanOptions>& opts = *scan_opts_;
+    std::vector<ScanOptions> subset;
+    subset.reserve(opts.size());
+    // We first find the first scan_opts start key that is in the file.
+    for (size_t i = 0; i < opts.size(); i++) {
+      const FdWithKeyRange& cur_file = flevel_->files[new_file_index];
+      auto target = opts[i].range.start.AsPtr();
+      if (user_comparator_.Compare(*target,
+                                   ExtractUserKey(cur_file.largest_key)) <= 0 &&
+          user_comparator_.Compare(
+              *target, ExtractUserKey(cur_file.smallest_key)) >= 0) {
+        subset.push_back(opts[i]);
```

Comment:
One tricky case to handle is start < smallest_key < limit < largest_key. In that case, LevelIterator will call SeekToFirst() when it moves to this file, so you may have to do `subset.push_back(ScanOptions(OptSlice(), opts[i].limit))`. This needs to be coordinated with @cbi42 and #13778 , since  BlockBasedTableIterator assumes we will always Seek to the start key in ScanOptions.